zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+    /// blocks that are in the chain, and `Err` for side-chain blocks.
+    ///
+    /// Does nothing if verification has finished.
+    ///
+    /// Check that the current checkpoint range is complete before calling this
+    /// function.
+    fn submit_current_checkpoint_range(&mut self) {
+        // If checkpoint verification has finished, there should be no queued blocks.
+        if self.current_checkpoint_range.is_none() {
+            return;
+        }
+
+        let next_checkpoint_height = self.get_next_checkpoint_height().unwrap();
+        let current_range = self.current_checkpoint_range.unwrap();
+
+        // Verify all the blocks and discard all the bad blocks in the current range.
+        let mut expected_hash = self.get_next_checkpoint_hash().unwrap();
+        let qrange = self.queued.range_mut(current_range).rev();
+        for (_, qblocks) in qrange {
+            // Find a queued block at each height that is part of the hash chain
+            //
+            // There are two possible outcomes here:
+            //   - there is one block, and it matches the chain (the common case)
+            //   - there are multiple blocks, and at least one block matches the chain
+            //     (if there are duplicate blocks, one succeeds, and the others fail)
+
+            // The caller should check for a continuous chain.
+            debug_assert!(!qblocks.is_empty());
+
+            let mut next_parent_hash = None;
+            for qblock in qblocks.drain(..) {
+                if qblock.hash == expected_hash {
+                    if next_parent_hash == None {
+                        // The first valid block at the current height
+                        next_parent_hash = Some(qblock.block.header.previous_block_hash);
+                        // TODO(teor): These futures are sent in reverse order. Make sure
+                        // the overall verifier adds blocks to the state in height order.
+                        // Sending can fail, but there's nothing we can do about it.
+                        let _ = qblock.tx.send(Ok(qblock.hash));
+                    } else {
+                        // Reject duplicate blocks at the same height
+                        let _ = qblock.tx.send(Err(
+                            "duplicate valid blocks at this height, only one was chosen".into(),
+                        ));
+                    }
+                } else {
+                    // A bad block, that isn't part of the chain.
+                    let _ = qblock.tx.send(Err(
+                        "the block hash does not match the chained checkpoint hash".into(),
+                    ));
+                }
+            }
+            // Can't fail, we just checked it in the previous loop
+            expected_hash = next_parent_hash.unwrap();
+        }
+
+        // Double-check that all the block lists are empty
+        #[cfg(any(dev, test))]
```

Comment:
I don't think we should include this kind of conditionally compiled assertion logic. I doubt we have any perf reasons to disable this in `--release` builds and these kinds of conditionally compiled blocks and assertions are hazardous. We have to be careful to never put expressions with side-effects in any of these blocks or `debug_assert`s.

Reply:
Removed as part of the refactor in 162b75ae4d71855c5a34183506535ff183e4ed53